core/src/main/java/ai/timefold/solver/core/impl/move/streams/maybeapi/generic/provider/ListSwapMoveProvider.java

```diff
@@ -0,0 +1,65 @@
+package ai.timefold.solver.core.impl.move.streams.maybeapi.generic.provider;
+
+import java.util.Objects;
+
+import ai.timefold.solver.core.impl.move.streams.maybeapi.DataJoiners;
+import ai.timefold.solver.core.impl.move.streams.maybeapi.generic.move.ListSwapMove;
+import ai.timefold.solver.core.impl.move.streams.maybeapi.stream.MoveProducer;
+import ai.timefold.solver.core.impl.move.streams.maybeapi.stream.MoveProvider;
+import ai.timefold.solver.core.impl.move.streams.maybeapi.stream.MoveStreamFactory;
+import ai.timefold.solver.core.preview.api.domain.metamodel.PlanningListVariableMetaModel;
+import ai.timefold.solver.core.preview.api.domain.metamodel.PositionInList;
+import ai.timefold.solver.core.preview.api.move.SolutionView;
+
+import org.jspecify.annotations.NullMarked;
+
+@NullMarked
+public class ListSwapMoveProvider<Solution_, Entity_, Value_>
+        implements MoveProvider<Solution_> {
+
+    private final PlanningListVariableMetaModel<Solution_, Entity_, Value_> variableMetaModel;
+
+    public ListSwapMoveProvider(PlanningListVariableMetaModel<Solution_, Entity_, Value_> variableMetaModel) {
+        this.variableMetaModel = Objects.requireNonNull(variableMetaModel);
+    }
+
+    @Override
+    public MoveProducer<Solution_> apply(MoveStreamFactory<Solution_> moveStreamFactory) {
+        var assignedValueStream = moveStreamFactory.enumerate(variableMetaModel.type(), false)
+                .filter((solutionView,
+                        value) -> solutionView.getPositionOf(variableMetaModel, value) instanceof PositionInList);
```

Comment:
We need to verify if the position is assigned in other places. What do you think about updating the `ElementPosition` contract to include a new method called `isAssigned`? This approach seems preferable to casting.